File: sdk/rust/src/artifact/rustc.rs
use crate::{
    artifact::{rust_toolchain, step, system, Artifact, ArtifactSource},
    context::ConfigContext,
};
use anyhow::Result;

/// Build-target `rustc` component of the Rust toolchain.
#[derive(Default)]
pub struct Rustc {}

impl Rustc {
    /// Creates a new `rustc` builder.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Builds the `rustc` artifact for the context's target system.
    ///
    /// # Errors
    ///
    /// Returns an error if the target system has no known Rust toolchain release, or if
    /// registering the source or artifact with the build context fails.
    pub async fn build(self, context: &mut ConfigContext) -> Result<String> {
        let name = "rustc";
        let system = context.get_system();

        let source_target = rust_toolchain::target(system)?;
        let source_version = rust_toolchain::version();
        let source_path = format!(
            "https://sdk.vorpal.build/source/{name}-{source_version}-{source_target}.tar.gz"
        );

        let source = ArtifactSource::new(name, source_path.as_str()).build();

        let step_script = format!("cp -pr \"./source/{name}/{name}-{source_version}-{source_target}/{name}/.\" \"$VORPAL_OUTPUT\"");
        let steps = vec![step::shell(context, &[], &[], step_script, &[]).await?];

        Artifact::new(name, steps, system::SYSTEMS)
            .with_sources(vec![source])
            .build(context)
            .await
    }
}

File: sdk/typescript/src/artifact/rustc.ts
import type { ConfigContext } from "../context.js";
import { Artifact, ArtifactSource } from "../artifact.js";
import { shell } from "./step.js";
import { RUST_TOOLCHAIN_VERSION, rustToolchainTarget } from "./rust_toolchain.js";
import { SYSTEMS } from "../system.js";

/**
 * Builder for the Rustc artifact.
 *
 * Mirrors Rust `Rustc` struct in `sdk/rust/src/artifact/rustc.rs`.
 * Downloads and extracts the Rust compiler from static.rust-lang.org.
 */
export class Rustc {
  async build(context: ConfigContext): Promise<string> {
    const name = "rustc";
    const system = context.getSystem();

    const sourceTarget = rustToolchainTarget(system);
    const sourceVersion = RUST_TOOLCHAIN_VERSION;
    const sourcePath = `https://sdk.vorpal.build/source/${name}-${sourceVersion}-${sourceTarget}.tar.gz`;

    const source = new ArtifactSource(name, sourcePath).build();

    const stepScript = `cp -pr "./source/${name}/${name}-${sourceVersion}-${sourceTarget}/${name}/." "$VORPAL_OUTPUT"`;
    const steps = [await shell(context, [], [], stepScript, [])];

    return new Artifact(name, steps, SYSTEMS)
      .withSources([source])
      .build(context);
  }
}
